### ISSUE TITLE
proper formatting of RTL/BiDi messages sent from k9

### ISSUE THREAD

**Issue by dovf** [thundernest/k-9#6520]
### Checklist

- [X] I have used the search function to see if someone else has already submitted the same feature request.
- [X] I will describe the problem with as much detail as possible.
- [X] This issue only contains a request for one single feature, **not** multiple (related) features.

### App version

6.400

### Problem you are trying to solve

I'd like to be able to send messages containing RTL/BiDi text, which will be rendered correctly by the receiver. Currently, when I send such messages, they render correctly only in k9 itself (and that only under specific circumstances), but not, e.g., in Gmail.

It's actually a bit complicated, so bear with me...

RTL languages (such as Hebrew / Arabic / Persian) are written right-to-left -- this refers both to the fact that the text is aligned to the right rather than to the left, as well as the fact that the second letter of a word is written to the left of  the first letter, and the second word is written to the left of the first word, etc. When mixing RTL and LTR text in the same paragraph, things are even more complicated. Proper rendering of such text is done by the BiDi (bidirectional) algorithm. Specifically (and simplifying a bit), part of the algorithm is that the paragraph direction is determined by the language of the first character.

In the following examples, "אבג" is the first logical word, with its first letter being "א" (Hebrew letter Aleph). The second word is "abc", and the third word is "דהו". As explained above, since the first character is in Hebrew (an RTL language) , the entire paragraph should be aligned to the right.

When *composing* the message in k9, it renders correctly:
![t1-compose-k9](https://user-images.githubusercontent.com/830975/206689673-3fb50c02-a2fa-4901-ba5e-e71710f4dff6.png)
Note that the first word (אבג) is rightmost, followed (moving from right-to-left) by the second word (abc) and then the third (דהו). Also, the entire paragraph is aligned to the right.

However, when *viewing* the received message in k9, it does not display correctly:
![t1-view-k9](https://user-images.githubusercontent.com/830975/206691308-8267ae32-6978-4d24-bb0f-687384b19652.png)
Not only is the entire text aligned to the left, but also the first word appears left-most, rather than right-most.

The received message is also rendered incorrectly when viewed in gmail:
![t1-view-gmail](https://user-images.githubusercontent.com/830975/206691954-737dc370-dc69-4e3e-b617-d73606a2afcd.png)

On the other hand, when trying the same test from Gmail, it displays correctly all along -- displays correctly while composing it in gmail:
![t3-compose-gmail](https://user-images.githubusercontent.com/830975/206692209-9af61a72-3fb8-4f4b-a428-7a2b724619f0.png)
renders correctly after being received in gmail:
![t3-view-gmail](https://user-images.githubusercontent.com/830975/206692276-a836479c-d78e-4890-a746-c6ee684d3467.png)
and also renders correctly when viewed in k9:
![t3-view-k9](https://user-images.githubusercontent.com/830975/206692360-294572a6-ccef-4305-82a6-87359d1b869b.png)

Note that both messages are composed in HTML/RTF format. However, when inspecting the messages, it can be seen that the one composed by gmail conatins html "dir" attributes (showing only the text/html part of the message):
```
Content-Type: text/html; charset="UTF-8"
Content-Transfer-Encoding: quoted-printable

<div dir=3D"ltr"><div dir=3D"rtl"><span class=3D"gmail_default" style=3D"fo=
nt-family:tahoma,sans-serif;font-size:small">=D7=90=D7=91=D7=92 abc =D7=93=
=D7=94=D7=95<br></span></div></div>
```
whereas the one composed by k9 does not:
```
Content-Type: text/html; charset=utf-8
Content-Transfer-Encoding: quoted-printable

<!DOCTYPE html><html><body>=D7=90=D7=91=D7=92 abc =D7=93=D7=94=D7=95</body>=
</html>
```

As I started looking into whether this issue has already been reported, I came across a bunch of issues which seemingly *were* related -- especially #3391 and the associated PR #4841 -- however, although they are marked as fixed, clearly a problem still persists.

So continuing to investigate, I noticed that the above fix is applying the "dir" attribute to a `<pre>` element. So I tried the same tests with plaintext-formatted messages, with the following results: while composing the message in k9, all is good:
![t2-compose-k9](https://user-images.githubusercontent.com/830975/206695560-ca27c8bc-b2c0-43d2-82cf-4a186d3740a1.png)
And now, also when viewing the received message in k9, it is rendered correctly!
![t2-view-k9](https://user-images.githubusercontent.com/830975/206695674-efcee0a8-d44f-41b4-bb56-8d9de01a5c0e.png)
However, when viewed in gmail it still renders incorrectly:
![t2-view-gmail](https://user-images.githubusercontent.com/830975/206695783-0da9b0f3-3f4b-485e-94f9-58911dbc536c.png)

The sent message is, truly, plaintext, so no formatting instructions can be applied to it. It is rendered correctly by k9 because the above-mentioned fix forces `<pre dir="auto"...` onto the text while rendering. However, gmail (and, seemingly, most other mail clients) do not do that, and so it is rendered incorrectly by them. One can argue about which approach is more correct, but as long as most of the world is not yet using k9, it means that k9 users still have a problem communicating with non-k9 users...

BTW, gmail approaches the issue differently (I'll spare the screenshots for this): when composing in plaintext mode it does not apply any bidi processing even during the composition -- and if the user tried to manually switch the text direction (the formatting UI in gmail allows that) it immediately switches to RTF format rather than plaintext.

### Suggested solution

The solution, I *think*, is to surround outgoing HTML-formatted messages with a `<div dir="auto"...` *as part of the sent content itself*.

I realize that part of the problem here is that k9 doesn't yet provide a rich-formatting UI, so such formatting can't be applied explicitly by the user. But until such a time the above behavior would seem to be correct (and innocuous). And in fact, even once a formatting UI is provided, it probably would *still* be correct to automatically apply such a "dir" attribute, unless explicitly overridden by the explicit formatting.

Not only would this solve the interoperability problem between k9 and non-k9 users, but, as observed, currently even within k9 itself, HTML-formatted message are rendered correctly while composing, but rendered *incorrectly* when viewed. The suggested fix would align the text as being rendered while composing with what would actually be viewed later on by k9 itself.

Thanks!

**Comment by dovf** [comment]
Rereading now the original issue #3391, I just wanted to add that the discussion there got very close to the issue I'm describing here -- specifically, question 3 and the response to it (https://github.com/thundernest/k-9/issues/3391#issuecomment-646583963) are correct, in that "dir" should not be added by k9 to *viewed* html content -- as stated there, it is up to the sender to apply the correct formatting to the original message. But that's precisely what k9 itself is not doing when sending such HTML-formatted content, and which the fix I'm proposing here would address...

**Comment by cketti** [comment]
Pull request #6522 should fix the issue. If you can build the app yourself, I'd appreciate some early feedback. If not, please report back after testing the next beta version (K-9 Mail 6.500) which should include the fix.

**Comment by dovf** [comment]
Thank you so much for the very prompt response!

I've tested it, and it's much better this way, thank you!

However, while playing around with it, I now see that the fix I proposed is still a bit simplistic: as it now stands, the *entire message* receives the direction of the first line, even if the message is composed of multiple lines (or paragraphs, really), each of which has its own direction. (The BiDi algorithm seems to treat each line (i.e., following a newline?) separately). So we still get situations where some paragraphs of the text (those whose language is of a different direction than the direction of the first paragraph) are garbled. 

A solution which seems to work (but I'm not sure that it's very pretty, nor whether it's easy to implement) is to replace each `<br>` with a `</div><div dir="auto">` (i.e. close the previous `div`, and open a new one which again chooses the direction automatically; all this *in addition* to the changes you already made). That seems to produce the correct effect -- the new `div` produces a line-break like `<br>`, and less than a `<p>` would, for example. (As rendered both by firefox and chrome.) But it doesn't feel like a very "correct" solution...

(Although... now that I'm looking to see what gmail does, take a look at this message produced by gmail -- without even mixing languages! -- simply by typing "abc" "def" "ghi" each on a new line:
```
<div dir=3D"ltr"><div class=3D"gmail_default" style=3D"font-family:tahoma,s=
ans-serif;font-size:small">abc</div><div class=3D"gmail_default" style=3D"f=
ont-family:tahoma,sans-serif;font-size:small">def</div><div class=3D"gmail_=
default" style=3D"font-family:tahoma,sans-serif;font-size:small">ghi<br></d=
iv></div>
```
So maybe it's not such a crazy suggestion...?)

But regardless, the current fix is already a big improvement, I think, and I would like to see it go in. (The only case where it would make things worse is the case where most of the message is in a different direction than the first line...). And then we can continue thinking about the next stage...

Thanks again for the responsiveness!
